File: projects/ntlm2/fuzz_auth.py
#!/usr/bin/python3
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import atheris
import sys
import socket
import time
from threading import Thread
with atheris.instrument_imports():
    import requests
    from requests.exceptions import InvalidURL,ConnectionError
    from requests_ntlm2 import HttpNtlmAuth

class ServerThread(Thread):
    def __init__(self, fdp):
        self.fdp = fdp

        self.s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.s.bind(("127.0.0.1", 8001))
        self.s.listen(1)

        Thread.__init__(self)

    def run(self):
        conn, addr = self.s.accept()
        conn.recv(self.fdp.ConsumeIntInRange(1024,2048))
        conn.send(self.fdp.ConsumeBytes(2048))
        time.sleep(0.005)
        conn.close()
        self.s.shutdown(1)
        self.s.close()
        time.sleep(0.01)

def TestInput(data):
    fdp = atheris.FuzzedDataProvider(data)

    t1 = ServerThread(fdp)
    t1.start()

    session = requests.Session()
    auth = HttpNtlmAuth(fdp.ConsumeString(50),fdp.ConsumeString(10))
    session.auth = HttpNtlmAuth(fdp.ConsumeString(50),fdp.ConsumeString(10))

    try:
       requests.get('http://localhost:8001/%s'%fdp.ConsumeString(20),auth=auth)
       session.get('http://localhost:8001/%s'%fdp.ConsumeString(20))
    except (InvalidURL,ConnectionError) as e:
       pass

    t1.join()

def main(): 
    atheris.Setup(sys.argv, TestInput, enable_python_coverage=True)
    atheris.instrument_all()
    atheris.Fuzz()

if __name__ == "__main__":
  main()

File: projects/dnsmasq/fuzz_auth.c
/* Copyright 2021 Google LLC
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
      http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include "fuzz_header.h"

/* 
 * Targets answer_auth
 */
void FuzzAuth(const uint8_t **data2, size_t *size2) {
  const uint8_t *data = *data2;
  size_t size = *size2;
  
  int i1 = get_int(&data, &size);
  int i2 = get_int(&data, &size);
  int i3 = get_int(&data, &size);

  if (size > (sizeof(struct dns_header) +50)) {
    char *new_data = malloc(size+1);
    memset(new_data, 0, size);
    memcpy(new_data, data, size);
    new_data[size] = '\0';
    pointer_arr[pointer_idx++] = (void*)new_data;

    time_t now;
		union mysockaddr peer_addr;
		answer_auth((struct dns_header *)new_data, new_data + size, size, now, &peer_addr, i1, i2, i3);
  }
}

/*
 * Fuzzer entrypoint.
 */ 
int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
  daemon = NULL;
  if (size < 1) {
    return 0;
  }

  // Initialize mini garbage collector
  gb_init();

  // Get a value we can use to decide which target to hit.
  int i = (int)data[0];
  data += 1;
  size -= 1;

  int succ = init_daemon(&data, &size);

  if (succ == 0) {
    cache_init();
    blockdata_init();

		FuzzAuth(&data, &size);

    cache_start_insert();
    fuzz_blockdata_cleanup();
  }

  // Free data in mini garbage collector.
  gb_cleanup();

  return 0;
}
